File: 22.generate-parentheses.js
/*
 * @lc app=leetcode id=22 lang=javascript
 *
 * [22] Generate Parentheses
 */

// @lc code=start
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
  let res = [];

  const helper = (leftP, rightP, str) => {
    if (leftP > rightP) return;
    if (!leftP && !rightP) res.push(str);
    if (leftP > 0) helper(leftP - 1, rightP, str + '(');
    if (rightP > 0) helper(leftP, rightP - 1, str + ')');
  };
  helper(n, n, '');
  return res;
};

// @lc code=end

File: 22.generate-parentheses.cpp
/*
 * @lc app=leetcode id=22 lang=cpp
 *
 * [22] Generate Parentheses
 */

// @lc code=start
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;

        helper(n, n, "", res);

        return res;
    }

    void helper(int leftParenthesis, int rightParenthesis, string str, vector<string>& res) {

        if (leftParenthesis > rightParenthesis)
            return;
        if (leftParenthesis == 0 && rightParenthesis == 0)
            res.push_back(str);
        if (leftParenthesis > 0)
            helper(leftParenthesis - 1, rightParenthesis, str + "(", res);
        if (rightParenthesis > 0)
            helper(leftParenthesis, rightParenthesis - 1, str + ")", res);
    }
};
// @lc code=end

